### PR TITLE
[sync part1] bucket: Improved bucket flag handling when no store is specifed; simplified factory and bucket commands code.

### PR DESCRIPTION
**Pull request by bwplotka**
This is prerequsite for https://github.com/improbable-eng/thanos/pull/731

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ We use *breaking* word for marking changes that are not backward compatible (rel
 
 ### Fixed
 
+- [#396](https://github.com/improbable-eng/thanos/issues/396) - Fixed sidecar missing proxying samples if Prometheus result for single series was longer than 2^16
```
Comment (bwplotka) [inline CHANGELOG.md]: sorry for unrelated change, but that was missed from last changeset. Can factor out.

File: cmd/thanos/flags.go [hunk 2]
```diff
@@ -157,19 +159,27 @@ func (p *pathOrContent) Content() ([]byte, error) {
 		return []byte(*p.content), nil
 	}
 
+	if p.required {
```
Comment (improbable-ludwik) [inline cmd/thanos/flags.go]: This will error on an empty file

File: cmd/thanos/flags.go [hunk 3]
```diff
@@ -134,21 +134,23 @@ func modelDuration(flags *kingpin.FlagClause) *model.Duration {
...
+		return nil, errors.Errorf("Both %s and %s flags set.", p.fileFlagName, p.contentFlagName)
 	}
 
 	if len(*p.path) > 0 {
 		c, err := ioutil.ReadFile(*p.path)
 		if err != nil {
-			return nil, errors.Wrap(err, fmt.Sprintf("loading YAML file %s for %s", *p.path, p.name))
+			return nil, errors.Wrapf(err, "loading YAML file %s for %s", *p.path, p.fileFlagName)
 		}
 		return c, nil
```
Comment (improbable-ludwik) [inline cmd/thanos/flags.go]: let be consistent between returning error and files being empty.

File: pkg/objstore/client/factory.go [hunk 4]
```diff
@@ -33,16 +33,10 @@ type BucketConfig struct {
 	Config interface{} `yaml:"config"`
 }
 
-var ErrNotFound = errors.New("not found bucket")
-
 // NewBucket initializes and returns new object storage clients.
 // NOTE: confContentYaml can contain secrets.
 func NewBucket(logger log.Logger, confContentYaml []byte, reg prometheus.Registerer, component string) (objstore.Bucket, error) {
 	level.Info(logger).Log("msg", "loading bucket configuration")
-	if len(confContentYaml) == 0 {
```
Comment (improbable-ludwik) [inline pkg/objstore/client/factory.go]: What's the worst that can happen if `NewBucket` is called with and empty config?

File: cmd/thanos/sidecar.go [hunk 5]
```diff
@@ -225,33 +227,36 @@ func runSidecar(
...
+	var uploads = true
+	if len(bucketConfig) == 0 {
 		level.Info(logger).Log("msg", "No supported bucket was configured, uploads will be disabled")
 		uploads = false
 	}
 
 	if uploads {
+		// The background shipper continuously scans the data directory and uploads
+		// new blocks to Google Cloud Storage or an S3-compatible storage service.
+		bkt, err := client.NewBucket(logger, bucketConfig, reg, component)
```
Comment (improbable-ludwik) [inline cmd/thanos/sidecar.go]: pkg/objstore/client/factory.go has the BucketConfig struct while the bucketConfig here is a  byte[] that we hope holds some yaml data that might match the BucketConfig struct. Consider renaming to  `bucketConfig` -> `confContentYaml`to match NewBucket implementation.

File: cmd/thanos/sidecar.go [hunk 6]
```diff
@@ -281,6 +286,20 @@ func runSidecar(
...
 }
 
+func isPrometheusDirAccesible(dir string) error {
+	const errMsg = "WAL file is not accessible. Does block shipper dir is a TSDB directory? If yes it is shared with TSDB?"
+	f, err := os.Stat(filepath.Join(dir, "WAL"))
+	if err != nil {
+		return errors.Wrap(err, errMsg)
+	}
+
+	if f.IsDir() {
```
Comment (domgreen) [inline cmd/thanos/sidecar.go]: `WAL` is a directory ... dont think we should error here.

File: cmd/thanos/bucket.go [hunk 7]
```diff
@@ -46,18 +46,24 @@ var (
...
-	verifyRepair := verify.Flag("repair", "Attempt to repair blocks for which issues were detected").
+	registerBucketVerify(m, cmd, name, objStoreConfig)
+	registerBucketLs(m, cmd, name, objStoreConfig)
+	registerBucketInspect(m, cmd, name, objStoreConfig)
+	return
+}
+
+func registerBucketVerify(m map[string]setupFunc, root *kingpin.CmdClause, name string, objStoreConfig *pathOrContent) {
+	cmd := root.Command("verify", "Verify all blocks in the bucket against specified issues")
+	objStoreBackupConfig := regCommonObjStoreFlags(cmd, "-backup", false)
```
Comment (domgreen) [inline cmd/thanos/bucket.go]: why do we need to use the backup bucket? Might be worth mentioning something in the description of the `verify` command.

File: cmd/thanos/bucket.go [hunk 8]
```diff
@@ -131,9 +140,11 @@ func registerBucket(m map[string]setupFunc, app *kingpin.Application, name strin
 
 		return v.Verify(ctx, idMatcher)
 	}
+}
 
-	ls := cmd.Command("ls", "List all blocks in the bucket")
-	lsOutput := ls.Flag("output", "Format in which to print each block's information. May be 'json' or custom template.").
+func registerBucketLs(m map[string]setupFunc, root *kingpin.CmdClause, name string, objStoreConfig *pathOrContent) {
+	cmd := root.Command("ls", "List all blocks in the bucket")
+	output := cmd.Flag("output", "Format in which to print each block's information. May be 'json' or custom template.").
```
Comment (domgreen) [inline cmd/thanos/bucket.go]: `Optional format in which to print each bloc's information. Options are 'json', 'wide' or a custom template.`?

File: docs/components/bucket.md [hunk 9]
```diff
@@ -50,12 +50,6 @@ Flags:
...
                            store configuration in YAML.
-      --objstore-backup.config-file=<bucket.config-yaml-path>  
-                           Path to YAML file that contains object store-backup
-                           configuration.
-      --objstore-backup.config=<bucket.config-yaml>  
-                           Alternative to 'objstore-backup.config-file' flag.
-                           Object store-backup configuration in YAML.
 
 Subcommands:
   bucket verify [<flags>]
```
Comment (domgreen) [inline docs/components/bucket.md]: Should add documentation for the `inspect` subcommand.

File: cmd/thanos/sidecar.go (hunk 6 above)
Comment (domgreen) [reply cmd/thanos/sidecar.go]: `if !f.IsDir() {`

File: cmd/thanos/sidecar.go [hunk 10]
```diff
@@ -281,6 +286,20 @@ func runSidecar(
 	return nil
 }
 
+func isPrometheusDirAccesible(dir string) error {
+	const errMsg = "WAL file is not accessible. Does block shipper dir is a TSDB directory? If yes it is shared with TSDB?"
```
[the remark this answers is not in the document]
Comment (domgreen) [reply cmd/thanos/sidecar.go]: `WAL directory is not accessible. Please ensure that the WAL / TSDB directory is accessible by the sidecar.`?

File: pkg/objstore/client/factory.go (hunk 4 above)
Comment (bwplotka) [reply pkg/objstore/client/factory.go]: Error from client for insufficient auth or no bucket name. (:

File: cmd/thanos/sidecar.go (hunk 6 above)
Comment (bwplotka) [reply cmd/thanos/sidecar.go]: Wut? I guess I might have fixed it in another PR.. will fix it here, sorry

**Comment by bwplotka** [comment]
Addressed comments. @domgreen @improbable-ludwik

File: docs/components/bucket.md (hunk 9 above)
Comment (domgreen) [reply docs/components/bucket.md]: I put a PR out to add what I meant ;) #781

**Pull request merged**